[MRG] Updated examples and tests that use scipy's lena

ogrisel opened the pull request:
This is a followup on #5920:
- squashed all the commits into one (and removed the annoying merged commit in the middle) to cleanup the history and make it easier to cherry-pick this fix to 0.17.1 later.
- fixed a dtype related error in the denoising example with recent numpy.

I have not fixed the circle ci configuration yet. Working on it next.

nelson-liu committed:
Updated examples and tests that use scipy's lena

scipy.misc.lena will be removed in scipy version 0.17.
This changes the example to use scipy.misc.face instead.

amueller commented on circle.yml:
```diff
@@ -1,25 +1,48 @@
   environment:
     # We need to set this variable to let python see the package installed through apt-get
-    PYTHONPATH: "/usr/lib/python2.7/dist-packages"
+    # PYTHONPATH: /usr/lib/python2.7/dist-packages
```
Why are these changes here?

ogrisel replied:
We have to use a more recent version of scipy to get the raccoon face in some examples so we cannot use the one from the system. I will soon push an amended commit that reworks circle ci that works on my account.

ogrisel commented:
@amueller I pushed some refactoring of the documentation build as part of this PR:

- upgraded circle ci env to use miniconda to build the doc (required to have the raccoon face in scipy)
- added a heuristic to skip the circle ci build in PRs that do not modify files under `doc/` or `examples/`
- as previously the push to the /dev/ section is only done from the master branch
- added the ability to skip the doc build in a PR by using the `[doc: skip]` marker in the last commit message
- added the ability to manually trigger the doc build in a PR using the  `[doc: build]` marker in the last commit message
- the circle ci doc is triggered by default on non-PR branches (e.g. push of a merge commit to master or cherrypicked fix commit pushed directly to 0.17.X)

I tested it on my circle ci account and it seems to work as expected.

ogrisel commented:
I think we can ignore the coverage decrease: the test_image.py module is tested on the "recent" matrix entry but it's just not taken into account by the coveralls stats. I tried to merge the reports but that is hard because the 3 builds do not have common filenames because of the specific build dirs.

ogrisel commented:
The doc build failed because of a random HTTP error in `quotes_historical_yahoo_ochl` that is not related to this PR.

Other than that the documentation looks nice:

https://circle-artifacts.com/gh/ogrisel/scikit-learn/65/artifacts/0/home/ubuntu/scikit-learn/doc/_build/html/stable/index.html

Here is the raccoon face segmentation example for instance:

https://circle-artifacts.com/gh/ogrisel/scikit-learn/65/artifacts/0/home/ubuntu/scikit-learn/doc/_build/html/stable/auto_examples/cluster/plot_face_segmentation.html#example-cluster-plot-face-segmentation-py

ogrisel commented:
@amueller @lesteve what do you think?

@AlexandreAbraham you might be interested in the `continuous_integration/circle/check_build_doc.py` script for the nilearn documentation.

amueller commented:
this looks really weird:
https://circle-artifacts.com/gh/ogrisel/scikit-learn/65/artifacts/0/home/ubuntu/scikit-learn/doc/_build/html/stable/auto_examples/cluster/plot_face_ward_segmentation.html

amueller commented:
the color quantization histogram looks different. what do the lines mean?

amueller commented:
It's really hard from the circleci output to see what the problem is :-/

ogrisel commented:
Indeed this example is weird. Can you reproduce when you run the example interactively on your machine?

ogrisel commented:
I merged #6040 so I will now rebase this on top of master.

ogrisel commented:
Ok I integrated the latex fix:

- https://circle-artifacts.com/gh/ogrisel/scikit-learn/66/artifacts/0/home/ubuntu/scikit-learn/doc/_build/html/stable/modules/sgd.html#mathematical-formulation

I also fixed the scrumbled raccoon face example:

- https://circle-artifacts.com/gh/ogrisel/scikit-learn/66/artifacts/0/home/ubuntu/scikit-learn/doc/_build/html/stable/auto_examples/cluster/plot_face_ward_segmentation.html#example-cluster-plot-face-ward-segmentation-py

It was a consequence of the use of [0-255] range uint8 dtype. I converted to float in the 0-1 range (and resized to 10% to make the example run much faster).

@amueller merge?

amueller commented:
LGTM.
